### ISSUE TITLE
tf.distribute.MultiWorkerMirroredStrategy with custom model.train_step

### ISSUE THREAD

**Issue by TZeng20** [tensorflow/tensorflow#58938]
### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow)**: yes
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 20.04
-   **TensorFlow version (use command below)**: 2.6
-   **Python version**: 3.8

### Describe the problem

I'd like to use tf.distribute.MultiWorkerMirroredStrategy with a custom model.train_step. How should the loss be calculated if using a custom loss? As an example, let's say my custom loss is tf.keras.losses.SparseCategoricalCrossentropy. Should the `reduction` be set to tf.keras.losses.Reduction.NONE (like below)? Or do I need to set the reduction to tf.keras.losses.Reduction.SUM and divide by the global batch size?

This problem is kind of in between keras.fit and writing custom distributed training loop. So not really sure how tensorflow will aggregate the gradients when calling model.fit. Just asking here because there doesn't seem to be documentation on this.
Attached a reproducible example below

### Source code / logs
```
import os
import tensorflow as tf
import numpy as np
import json

def mnist_dataset():
    '''
    Load mnist dataset. No batching
    '''
    (x_train, y_train), _ = tf.keras.datasets.mnist.load_data()
    x_train = x_train / np.float32(255)
    y_train = y_train.astype(np.int64)
    train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train)).shuffle(60000)
    return train_dataset

class CustomModel(tf.keras.Model):
    def __init__(self, model):
        super(CustomModel, self).__init__()
        self.model = model
        self.loss_tracker= tf.keras.metrics.Mean(name='loss')
        
    def train_step(self, data):
        x, y = data
        with tf.GradientTape() as tape:
            # Forward pass
            y_pred = self.model(x, training=True)
            # Compute our own loss. Shape (batch_size,)
            loss = tf.keras.losses.SparseCategoricalCrossentropy(
                from_logits = True, reduction=tf.keras.losses.Reduction.NONE)(y, y_pred)
           
        # Compute gradients
        trainable_vars = self.model.trainable_variables
        gradients = tape.gradient(loss, trainable_vars)
        # Update weights
        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
        # Update metrics
        self.loss_tracker.update_state(loss)
        self.compiled_metrics.update_state(y, y_pred)
        metrics = {m.name: m.result() for m in self.metrics}
        return metrics

def build_and_compile_cnn_model():
    model = tf.keras.Sequential([
        tf.keras.layers.InputLayer(input_shape=(28, 28)),
        tf.keras.layers.Reshape(target_shape=(28, 28, 1)),
        tf.keras.layers.Conv2D(32, 3, activation='relu'),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.Dense(10)])
    model = CustomModel(model = model)
    model.compile(optimizer=tf.keras.optimizers.SGD(learning_rate=0.001))
    return model

# TRAIN
per_worker_batch_size = 32
num_workers = len(tf_config['cluster']['worker'])
global_batch_size = per_worker_batch_size * num_workers
multi_worker_dataset = mnist_dataset().batch(global_batch_size)

with strategy.scope():
  # Model building/compiling need to be within `strategy.scope()`.
  multi_worker_model = build_and_compile_cnn_model()

multi_worker_model.fit(multi_worker_dataset, epochs=3, steps_per_epoch=70)
```

**Comment by TZeng20** [comment]
Sure. Updated with a short reproducible example

**Comment by SuryanarayanaY** [comment]
Hi @TZeng20 ,
Could you please refer the Source [here](https://www.tensorflow.org/tutorials/distribute/multi_worker_with_ctl#define_a_custom_training_loop_and_train_the_model) for Multi Worker Training and custom training loop. You need to create per_replica step function under `train_step()` like below. 

```
@tf.function
def train_step(iterator):
  """Training step function."""

  def step_fn(inputs):
    """Per-Replica step function."""
    x, y = inputs
    with tf.GradientTape() as tape:
      predictions = multi_worker_model(x, training=True)
      per_batch_loss = tf.keras.losses.SparseCategoricalCrossentropy(
          from_logits=True,
          reduction=tf.keras.losses.Reduction.NONE)(y, predictions)
      loss = tf.nn.compute_average_loss(
          per_batch_loss, global_batch_size=global_batch_size)

    grads = tape.gradient(loss, multi_worker_model.trainable_variables)
    optimizer.apply_gradients(
        zip(grads, multi_worker_model.trainable_variables))
    train_accuracy.update_state(y, predictions)
    return loss

  per_replica_losses = strategy.run(step_fn, args=(next(iterator),))
  return strategy.reduce(
      tf.distribute.ReduceOp.SUM, per_replica_losses, axis=None)
```
Please refer to above source and let us know if still having any problems.

Thank you !

**Comment by TZeng20** [comment]
Hi @SuryanarayanaY, 
Thanks for your reply, however that is not what I'm looking for. I was looking to use the default tf.keras.Model.fit with a custom train step, like [here](https://www.tensorflow.org/guide/keras/customizing_what_happens_in_fit). 

The code you have attached does not use the default model.fit method. So I may not have access to `global_batch_size`, and also not sure if model.fit uses tf.distribute.ReduceOp.SUM to aggregate gradients of each replica.

**Comment by SuryanarayanaY** [comment]
Hi @TZeng20 ,

If using tf.keras.losses classes, the loss reduction needs to be explicitly specified to be one of NONE or SUM.Please refer to one of the code snippet below.
```
with strategy.scope():
  # Set reduction to `NONE` so you can do the reduction afterwards and divide by
  # global batch size.
  loss_object = tf.keras.losses.SparseCategoricalCrossentropy(
      from_logits=True,
      reduction=tf.keras.losses.Reduction.NONE)
  def compute_loss(labels, predictions, model_losses):
    per_example_loss = loss_object(labels, predictions)
    loss = tf.nn.compute_average_loss(per_example_loss,
                                      global_batch_size=GLOBAL_BATCH_SIZE)
    if model_losses:
      loss += tf.nn.scale_regularization_loss(tf.add_n(model_losses))
    return loss
```

For more details please refer to attached documentation [here](https://www.tensorflow.org/tutorials/distribute/custom_training).

Please refer the attached source and let us know if it helps.Please come back if the issue still persists.

Thankyou!

**Comment by SuryanarayanaY** [comment]
@TZeng20 ,

Global batch size can get as below.

`GLOBAL_BATCH_SIZE = BATCH_SIZE_PER_REPLICA * strategy.num_replicas_in_sync`

BATCH_SIZE_PER_REPLICA = Your own defined batch size
strategy = Strategy you are using
num_replicas_in_sync = attribute of strategy(inbuilt)

**Comment by TZeng20** [comment]
Is the following code the correct way to override the training step function of the Model class? Please note that I am subclassing tf.keras.Model. In the code you have provided `train_step` is a global function, but mine is an instance method. So I would not have access to the global variables `global_batch_size` and `strategy`

```
class CustomModel(tf.keras.Model):
    def __init__(self, model):
        super(CustomModel, self).__init__()
        self.model = model
        self.loss_tracker= tf.keras.metrics.Mean(name='loss')
        
    def train_step(self, data):
        x, y = data
        replica_context = tf.distribute.get_replica_context()
        tf.print("Replica id: ", replica_context.replica_id_in_sync_group, 
                    " of ", replica_context.num_replicas_in_sync)
        GLOBAL_BATCH_SIZE = replica_context.num_replicas_in_sync*tf.shape(y)[0]
        with tf.GradientTape() as tape:
            # Forward pass
            y_pred = self.model(x, training=True)
            # Compute our own loss. Shape (batch_size,)
            loss = tf.keras.losses.SparseCategoricalCrossentropy(
                from_logits = True, reduction=tf.keras.losses.Reduction.NONE)(y, y_pred)
            loss = tf.nn.compute_average_loss(loss, global_batch_size=GLOBAL_BATCH_SIZE)
            
        # Compute gradients
        trainable_vars = self.model.trainable_variables
        gradients = tape.gradient(loss, trainable_vars)
        # Update weights
        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
        # Update metrics
        self.loss_tracker.update_state(loss)
        self.compiled_metrics.update_state(y, y_pred)
        metrics = {m.name: m.result() for m in self.metrics}
        return metrics
```

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

Could you please refer the doc from [here](https://www.tensorflow.org/tutorials/distribute/multi_worker_with_ctl#define_a_custom_training_loop_and_train_the_model) for Multi Worker Training and custom training loop. Also try  to create per_replica step function under train_step() as mentioned.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
